### PR TITLE
TRPO: correct the params positions of diagonal gaussian kl function.

### PR DESCRIPTION
**Pull request by 2foil**
## Background

The KL divergence (Kullback–Leibler divergence) is a measure of how one probability distribution is different from a second.

And the KL divergence is asymmetric, which means

To convey the fundamental asymmetry of the KL divergence,  is described as the divergence of P from Q or the divergence from Q to P.

## Spinup code
The `diagonal_gaussian_kl` function implemented in spinup, calculates the KL divergence  from N1 to N0:

```python3
# spinningup/spinup/algos/trpo/core.py

def diagonal_gaussian_kl(mu0, log_std0, mu1, log_std1):
    """
    tf symbol for mean KL divergence between two batches of diagonal gaussian distributions,
    where distributions are specified by means and log stds.
    (https://en.wikipedia.org/wiki/Kullback-Leibler_divergence#Multivariate_normal_distributions)
    """
    var0, var1 = tf.exp(2 * log_std0), tf.exp(2 * log_std1)
    pre_sum = 0.5*(((mu1- mu0)**2 + var0)/(var1 + EPS) - 1) +  log_std1 - log_std0
    all_kls = tf.reduce_sum(pre_sum, axis=1)
    return tf.reduce_mean(all_kls)
```
which equals to:

Why I use the term ***from N1 to N0*** here is because of the fundamental asymmetry of KL divergence, which means

When invoking the `diagonal_gaussian_kl` function at the `mlp_gaussian_policy` function,
```python3
# spinningup/spinup/algos/trpo/core.py

def mlp_gaussian_policy(x, a, hidden_sizes, activation, output_activation, action_space):
    ...
    d_kl = diagonal_gaussian_kl(mu, log_std, old_mu_ph, old_log_std_ph)
    ...
```
the first two params represent the new policy and the last two params represent the old policy.
So the `d_kl` is the symbol of KL divergence ***from old policy to new policy***, which means:

## Problem

In the original TRPO paper, the trust region constraint is defined as:

where  represents the KL divergence ***from new policy to old policy***. But in the code, the `d_kl` symbol calculates the KL divergence ***from old policy to new policy***, which is incorrect.
So the params to invoke the `diagonal_gaussian_kl` function is in a wrong order, we need correct it. 
Another equation in the original TRPO paper also confirms me in that the param order is wrong:

## Bug fix 
There are two options to fix this bug:
1. Change the param order when invoking the `diagonal_gaussian_kl` function:
We can replace  
`d_kl = diagonal_gaussian_kl(mu, log_std, old_mu_ph, old_log_std_ph)` 
with 
`d_kl = diagonal_gaussian_kl(old_mu_ph, old_log_std_ph, mu, log_std)`.

2. Keep the param order not changed, change the inside logic of `diagonal_gaussian_kl` function:
We can change the implementation of `diagonal_gaussian_kl` to get the correct KL divergence.
The new `diagonal_gaussian_kl` function is:
```python3
# spinningup/spinup/algos/trpo/core.py

def diagonal_gaussian_kl(mu0, log_std0, mu1, log_std1):
    """
    tf symbol for mean KL divergence between two batches of diagonal gaussian distributions,
    where distributions are specified by means and log stds.
    (https://en.wikipedia.org/wiki/Kullback-Leibler_divergence#Multivariate_normal_distributions)
    """
    var0, var1 = tf.exp(2 * log_std0), tf.exp(2 * log_std1)
    pre_sum = 0.5*(((mu0- mu1)**2 + var1)/(var0 + EPS) - 1) +  log_std0 - log_std1
    all_kls = tf.reduce_sum(pre_sum, axis=1)
    return tf.reduce_mean(all_kls)
```
which equals to:

When invoking it with `d_kl = diagonal_gaussian_kl(mu, log_std, old_mu_ph, old_log_std_ph)`, the `d_kl` is the symbol of KL divergence ***from new policy to old policy***:

The first option is simple, so I take it. 

## Reference 
- [Kullback-Leibler divergence ](https://en.wikipedia.org/w/index.php?title=Kullback%E2%80%93Leibler_divergence#Examples)
- [Trust Region Policy Optimization](https://arxiv.org/abs/1502.05477)

### CODE REVIEW

**Comment by jachiam** [comment]
Hello! Indeed, you are correct that KL-divergence is not symmetric, and the original paper does things one way, and here we do things a slightly-different way. However, I counter-claim that this is not actually a bug---there is no error, from either a mathematical standpoint (in terms of the theoretical bounds on which TRPO is ostensibly based), or from an outcomes standpoint (in terms of the difference between using KL divergence in one direction or the other in TRPO).

From the mathematical standpoint: the theoretical bounds involved in TRPO in an earlier step use a symmetric divergence (the TV divergence), and only afterwards bound that by a KL divergence. The KL for either order can be used, so the choice of direction is arbitrary in theory.

From an outcomes standpoint: the TRPO computation depends on the KL divergence through its Hessian. It can be shown that the Hessian for KL, when evaluated with both arguments at the same distribution with the same parameterization (P=Q), is the same whether you were taking gradients with respect to the first distribution's parameters or the second distribution's parameters. Therefore, the parameter-space update direction you compute (H^{-1} g) is the same for either KL direction. The only place where a difference in calculation could occur for different choices of KL direction, then, is in the line search, which evaluates KL analytically (instead of with the Hessian). But the second-order approximation is good for small parameter distances, so we should expect that both KL directions will be approximately equal around the trust region, and so there will be virtually no difference in line search outcomes.

**Comment by 2foil** [comment]
@jachiam 
Thanks for your explanation. :)
